File: solutions/cpp/25326/main.cpp
#include<bits/stdc++.h>

using namespace std;

int cnt[3][3][3] = {};
vector<vector<string>> object = {
    {"kor", "eng", "math"},
    {"apple", "pear", "orange"},
    {"red", "blue", "green"}
};

int Found(int idx, string str) {
    for(int i = 0; i < 3; ++i) {
        if(object[idx][i] == str) return i;
    }
    return -1;
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(0);

    int N, M; cin >> N >> M;
    for(int i = 0; i < N; ++i) {
        string a, b, c; cin >> a >> b >> c;
        int idxa = Found(0, a);
        int idxb = Found(1, b);
        int idxc = Found(2, c);
        ++ cnt[idxa][idxb][idxc];
    }
    for(int i = 0; i < M; ++i) {
        string a, b, c; cin >> a >> b >> c;
        int idxa = Found(0, a);
        int idxb = Found(1, b);
        int idxc = Found(2, c);
        int ans = 0;
        for(int i = (idxa == -1 ? 0 : idxa); i <= (idxa == -1 ? 2 : idxa); ++i) {
            for(int j = (idxb == -1 ? 0 : idxb); j <= (idxb == -1 ? 2 : idxb); ++j) {
                for(int k = (idxc == -1 ? 0 : idxc); k <= (idxc == -1 ? 2 : idxc); ++k) {
                    ans += cnt[i][j][k];
                }
            }
        }
        cout << ans << '\n';
    }
    
    return 0;
}

File: solutions/python/25326/main.py
import sys
from typing import List


def input() -> str:
    return sys.stdin.readline().rstrip()


Objects: List[List[str]] = [
    ["kor", "eng", "math"],
    ["apple", "pear", "orange"],
    ["red", "blue", "green"],
]


def get_index(type: int, value: str) -> int:
    for i in range(3):
        if Objects[type][i] == value:
            return i
    return -1


def solution(
    N: int, M: int, preference: List[List[str]], query: List[List[str]]
) -> List[int]:
    cnt: List[List[List[int]]] = [
        [[0 for k in range(3)] for j in range(3)] for i in range(3)
    ]

    for x, y, z in preference:
        idx_x = get_index(0, x)
        idx_y = get_index(1, y)
        idx_z = get_index(2, z)
        cnt[idx_x][idx_y][idx_z] += 1

    answer: List[int] = []
    for x, y, z in query:
        idx_x = get_index(0, x)
        idx_y = get_index(1, y)
        idx_z = get_index(2, z)
        range_x = range(0, 3) if idx_x == -1 else [idx_x]
        range_y = range(0, 3) if idx_y == -1 else [idx_y]
        range_z = range(0, 3) if idx_z == -1 else [idx_z]
        ans = 0
        for i in range_x:
            for j in range_y:
                for k in range_z:
                    ans += cnt[i][j][k]
        answer.append(ans)
    return answer


if __name__ == "__main__":
    N, M = map(int, input().split())  # type: (int, int)
    preference: List[List[str]] = [input().split() for i in range(N)]
    query: List[List[str]] = [input().split() for i in range(M)]
    answer: List[int] = solution(N, M, preference, query)
    answer = "\n".join(map(str, answer))
    print(answer)
